### PR TITLE
Documenting Node and NodeGroup

### PR DESCRIPTION
**Pull request by leenlab2**
Added documentation for the functions and methods in `Node.js` and `NodeGroup.js`, as well as comments explaining certain sections of code.

There were some functions that I did not fully understand, so I did not add proper documentation to them but instead placed a template of one that we could update in the future.

### CODE REVIEW

File: js/components/graph/Node.js [hunk 1]
```diff
@@ -2,57 +2,99 @@ import PropTypes from "prop-types";
...
 
+/** React component class representing a Node on the graph */
 export default class Node extends React.Component {
+  /** Create a node */
   constructor(props) {
     super(props);
     var state = localStorage.getItem(this.props.JSON.id_);
+
+    /**
+     * Unselected states:
```
Comment (david-yz-liu) [inline js/components/graph/Node.js:13]: This is great documentation, but move it to the class documentation above.

File: js/components/graph/Node.js [hunk 2]
```diff
@@ -2,57 +2,99 @@ import PropTypes from "prop-types";
...
+     * 
+     * Selected states:
+     *  - active: all prerequisites are satisfied
+     *  - overridden: missing some prerequisites (will have a red border)
+     */
+
     if (this.props.editMode) {
       state = ""; // TODO: define what editMode is
     } else if (state === null) {
+      // sets state to takeable if it has no parent nodes (prereqs)
```
Comment (david-yz-liu) [inline js/components/graph/Node.js:25]: Remove this comment. It is redundant with the description of the states.

File: js/components/graph/Node.js [hunk 3]
```diff
@@ -2,57 +2,99 @@ import PropTypes from "prop-types";
...
       state = this.props.parents.length === 0 ? "takeable" : "inactive";
     }
+
```
Comment (david-yz-liu) [inline js/components/graph/Node.js:28]: Please revert all blank line changes. They aren't part of the scope of this pull request.

File: js/components/graph/Node.js [hunk 4]
```diff
@@ -2,57 +2,99 @@ import PropTypes from "prop-types";
...
     this.state = {
       status: state,
-      selected: ["active", "overridden"].indexOf(state) >= 0
+      selected: ["active", "overridden"].indexOf(state) >= 0 // checks if state means its selected
```
Comment (david-yz-liu) [inline js/components/graph/Node.js:31]: Remove this comment. Instead, it's better to add a note in the class documentation above that `selected` is a state.

File: js/components/graph/Node.js [hunk 5]
```diff
@@ -2,57 +2,99 @@ import PropTypes from "prop-types";
...
     };
   }
 
+  /**
+   * Checks whether this Node is selected
+   * @return {boolean}
+   */
   isSelected = () => {
+    // Hybrid nodes: the grey, smaller nodes representing another node farther way
```
Comment (david-yz-liu) [inline js/components/graph/Node.js:40]: Again, move this to the class documentation, as a separate part where you discuss what hybrid nodes are.

File: js/components/graph/Node.js [hunk 6]
```diff
@@ -2,57 +2,99 @@ import PropTypes from "prop-types";
...
+    // They can only be either active or inactive.
     if (this.props.hybrid) {
       return this.state.status === "active";
     } else {
       return this.state.selected;
     }
   }
 
+  /**
+   * Checks whether all prerequisite/ preceeding nodes for the current one are satisfied
```
Comment (david-yz-liu) [inline js/components/graph/Node.js:50]: Do not include spaces after `/`. Also, "preceeding" is spelled "preceding".

File: js/components/graph/Node.js [hunk 7]
```diff
@@ -2,57 +2,99 @@ import PropTypes from "prop-types";
...
-    var svg = this.props.svg;
+    var svg = this.props.svg;  // contains information about the whole graph
+
+    /**
+     * Recursively checks that preceeding nodes are selected
+     * @param  {string|Array} element Node(s)/ other on the graph
+     * @return {boolean}
+     */
     function isAllTrue(element) {
+      // base case: single node
```
Comment (david-yz-liu) [inline js/components/graph/Node.js:62]: Remove all inline comments in this function, and the one on line 73; they are just repeating what the code does, and in general we should not include such comments.

File: js/components/graph/Node.js [hunk 8]
```diff
@@ -2,57 +2,99 @@ import PropTypes from "prop-types";
...
   }
 
+  /**
+   * Update the state/status of a node (and its children/edges)
+   * @param  {boolean} recursive whether we should recurse on its children
+   */
   updateNode = recursive => {
     var newState;
+
+    // if the prereqs are satisfied, the node is takeable. Once selected it becomes active.
```
Comment (david-yz-liu) [inline js/components/graph/Node.js:89]: Remove the inline comments in this function (same comment as above).

File: js/components/graph/Node.js [hunk 9]
```diff
@@ -97,19 +143,23 @@ export default class Node extends React.Component {
...
     this.setState({ selected: !this.state.selected }, function() {
       this.updateNode();
     });
   }
 
+  /** Sets the status of all missing prerequisites to 'missing' */
   focusPrereqs = () => {
-    var svg = this.props.svg;
-    // Check if there are any missing prerequisites.
+    var svg = this.props.svg;  // contains information about the whole graph
```
Comment (david-yz-liu) [inline js/components/graph/Node.js:155]: Remove the inline comments

File: js/components/graph/Node.js [hunk 10]
```diff
@@ -139,10 +192,17 @@ export default class Node extends React.Component {
...
   }
 
+  /**
+   * Resets 'missing' nodes and edges to the previous statuses:
+   *  active, inactive, overridden, takeable
+   */
   unfocusPrereqs = () => {
-    var svg = this.props.svg;
-    this.updateNode(false);
+    var svg = this.props.svg;  // contains information about the whole graph
```
Comment (david-yz-liu) [inline js/components/graph/Node.js:200]: Remove the inline comments

File: js/components/graph/NodeGroup.js [hunk 11]
```diff
@@ -2,15 +2,19 @@ import PropTypes from "prop-types";
 import React from "react";
 import Node from "./Node";
 
+/** A React component class representing a group of Nodes on the graph */
 export default class NodeGroup extends React.Component {
+  /** Returns nodes to their original unselected state, with a status of "takeable" or "inactive" */
   reset = () => {
+    // Resets course nodes
```
Comment (david-yz-liu) [inline js/components/graph/NodeGroup.js:9]: Remove the inline comments

File: js/components/graph/NodeGroup.js [hunk 12]
```diff
@@ -43,17 +56,21 @@ export default class NodeGroup extends React.Component {
         {this.props.hybridsJSON.map(entry => {
           var childs = [];
           var outEdges = [];
+
+          // build childs and outEdges
```
Comment (david-yz-liu) [inline js/components/graph/NodeGroup.js:60]: Remove this comment

File: js/components/graph/NodeGroup.js [hunk 13]
```diff
@@ -43,17 +56,21 @@ export default class NodeGroup extends React.Component {
         {this.props.hybridsJSON.map(entry => {
           var childs = [];
           var outEdges = [];
+
```
Comment (david-yz-liu) [inline js/components/graph/NodeGroup.js:59]: Remove the blank line changes here.

**Review by david-yz-liu — COMMENTED** [review]
@leenlab2 a good start. I left a few different comments for you to address. My overall comment is that you should not be including inline comments to describe what code does, as the code already communicates this. (This is different from CS classes where you are explicitly learning the language. The standard for reading and understanding code is higher in a real-world project.)

A good rule of thumb for inline comments is that you should only include them to describe *why* a piece of code is written the way that it is, rather than *what* the code does.

File: js/components/graph/Node.js (hunk 7 above)
Comment (leenlab2) [reply js/components/graph/Node.js:62]: Should I remove the one next to var svg...?

File: js/components/graph/Node.js (hunk 8 above)
Comment (leenlab2) [reply js/components/graph/Node.js:89]: The one on line 107 was not added by me, should I keep it?

File: js/components/graph/Node.js (hunk 7 above)
Comment (david-yz-liu) [reply js/components/graph/Node.js:62]: Oh, yes that one too

File: js/components/graph/Node.js (hunk 8 above)
Comment (david-yz-liu) [reply js/components/graph/Node.js:89]: Good question! That's an example of a comment which isn't adding much, but could be improved to explain the *why* of the condition. So why don't you change it to a comment like "We can stop without updating the children when the selected state of this node hasn't changed and the original state wasn't "missing"."

File: js/components/graph/Node.js (hunk 9 above)
Comment (leenlab2) [reply js/components/graph/Node.js:155]: The comment on line 157 was also not added by me, should I improve or delete it?

Comment (david-yz-liu) [reply js/components/graph/Node.js:155]: Hmm, you can try to improve that one!

Comment (leenlab2) [reply js/components/graph/Node.js:155]: I decided to improve it to answer "why",  like we did to the other one

**Comment by leenlab2** [comment]
I applied some of your comments with these new commits. That includes:

- Removed blank line changes
- Removed unnecessary inline comments
- Made the class description more detailed, including describing the states, the status messages, and the types of nodes

Let me know what else needs fixing!

**Comment by david-yz-liu** [comment]
@leenlab2 great changes! You should leave the comments open so that I can see them when I review your changes.

**Pull request merged**